tensorboard/plugins/graph/graph_util.py

```diff
@@ -0,0 +1,136 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def keras_model_to_graph_def(keras_layer):
+  """Returns GraphDef representation of the model.
+
+  Note that it only supports models that implemented get_config().
+
+  Args:
+    keras_layer: A dict from Keras model.to_json()
+
+  Returns:
+    A GraphDef representation of layers.
+  """
+
+  def get_layers(keras_layer):
+    yield ('', keras_layer)
+    if keras_layer.get('config').get('layers'):
+      name_scope = keras_layer.get('config').get('name')
+      for layer in keras_layer.get('config').get('layers'):
+        for (sub_name_scope, sublayer) in get_layers(layer):
+          sub_name_scope = '%s/%s' % (
+              name_scope, sub_name_scope) if sub_name_scope else name_scope
+          yield (sub_name_scope, sublayer)
+
+  def scoped_name(name_scope, node_name):
+    if name_scope:
+      return '%s/%s' % (name_scope, node_name)
+    return node_name
+
+  model_inputs_to_layer = {}
+  model_name_to_output_layer = {}
+  g = GraphDef()
+
+  # Sequential model layers do not have "inbound_nodes".
+  # Must assume all nodes are connected linearly.
+  prev_node_name = None
+  for (name_scope, layer) in get_layers(keras_layer):
+    layer_config = layer.get('config')
+    node_name = scoped_name(name_scope, layer_config.get('name'))
+
+    if layer_config.get('layers'):
+      # Do not add models to GraphDef. Store information for models.
+      input_layers = layer_config.get('input_layers')
+      inbound_nodes = layer.get('inbound_nodes')
+      if input_layers and inbound_nodes:
```

Comment:
Consider adding a function that returns the type of parent and model here as an enumeration:

(parent_type, model_type) = get_parent_and_model_types(layer_config, layer)
...
if parent_type is FUNCTIONAL and model_type is FUNCTIONAL:
...
elif parent_type is FUNCTIONAL and model_type is SEQUENTIAL:
...

This will make this more readable IMO.